src/dataset.cpp

```diff
@@ -593,6 +593,87 @@ Dataset rebin(const Dataset &d, const Variable &newCoord) {
   return out;
 }
 
+Dataset histogram(const Variable &var, const Variable &coord) {
+  // TODO Is there are more generic way to find "histogrammable" data, not
+  // specific to (neutron) events? Something like Data::ValueVector, i.e., any
+  // data variable that contains a vector of values at each point?
+  const auto &events = var.get<const Data::Events>();
+  // TODO This way of handling events (and their units) as nested Dataset feels
+  // a bit unwieldy. Would it be a better option to store TOF (or any derived
+  // values) as simple vectors in Data::Events? There would be a separate
+  // Data::PulseTimes (and Data::EventWeights). This can then be of arbitrary
+  // type, unit conversion is reflected in the unit of Data::Events. The
+  // implementation of `histogram` would then also be simplified since we do not
+  // need to distinguish between Data::Tof, etc. (which we are anyway not doing
+  // currently).
+  dataset::expect::equals(events[0](Data::Tof{}).unit(), coord.unit());
+
+  // TODO Can we reuse some code for bin handling from DatasetView?
+  const auto binDim = coordDimension[coord.tag().value()];
+  const gsl::index nBin = coord.dimensions()[binDim] - 1;
+  Dimensions dims = var.dimensions();
+  if (dims.contains(binDim))
+    throw std::runtime_error(
+        "Data to histogram depends on histogram dimension.");
+  for (const auto &dim : coord.dimensions().labels()) {
+    if (dim != binDim) {
+      dataset::expect::dimensionMatches(dims, dim, coord.dimensions()[dim]);
+    }
+  }
+
+  dims.addInner(binDim, nBin);
+  const gsl::index nextEdgeOffset = coord.dimensions().offset(binDim);
+
+  Dataset hist;
+  hist.insert(coord);
+  hist.insert<Data::Value>(var.name(), dims);
+
+  auto counts = hist.get<Data::Value>(var.name());
+  gsl::index cur = 0;
+  // The helper `getView` allows us to ignore the tag of coord, as long as the
+  // underlying type is `double`.
+  // TODO Need to add a branch for the `float` case.
+  const auto edges = getView<double>(coord, dims);
+  auto edge = edges.begin();
+  for (const auto &eventList : events) {
```

Comment:
Each input eventlist looped over but single output histogram?

Reply:
`counts` is multi-dimensional, inner dimension is the bin-dimensions (the inner loop below), then it effectively jumps to the first bin of the next histogram.